// react-native-circular-chart/dist/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React, { useEffect, useRef, useMemo, useState, Fragment } from "react";
import { Text, View, Animated, StyleSheet, Easing, } from "react-native";
import { Svg, Path } from "react-native-svg";
import { Square } from "./packages/shape";
import { Arc, ViewBox } from "./packages/svg";
import { sum } from "./packages/array";
import { LinearInterpolation } from "./packages/math";
var AnimatedPath = Animated.createAnimatedComponent(Path);
export var DonutChart = function (_a) {
    var data = _a.data, containerWidth = _a.containerWidth, containerHeight = _a.containerHeight, radius = _a.radius, _b = _a.startAngle, startAngle = _b === void 0 ? -125 : _b, _c = _a.endAngle, endAngle = _c === void 0 ? startAngle * -1 : _c, _d = _a.strokeWidth, strokeWidth = _d === void 0 ? 10 : _d, _e = _a.type, type = _e === void 0 ? "round" : _e, _f = _a.animationType, animationType = _f === void 0 ? "slide" : _f, labelWrapperStyle = _a.labelWrapperStyle, labelValueStyle = _a.labelValueStyle, labelTitleStyle = _a.labelTitleStyle, containerStyle = _a.containerStyle;
    var donutItemListeners = [];
    var viewBox = new ViewBox({
        width: containerWidth,
        height: containerHeight,
    });
    var squareInCircle = new Square({ diameter: radius * 2 });
    var animateOpacity = useRef(new Animated.Value(0)).current;
    var animateContainerOpacity = useRef(new Animated.Value(0)).current;
    var animatedStrokeWidths = useRef(data.map(function () { return new Animated.Value(strokeWidth); })).current;
    var pathRefs = useRef([]);
    var animatedPaths = useRef([]).current;
    var _g = useState(data[0]), displayValue = _g[0], setDisplayValue = _g[1];
    // TODO:
    // remove WTF is this variable ?
    var _h = useState([]), rotationPaths = _h[0], setRotationPath = _h[1];
    var defaultInterpolateConfig = function () { return ({ inputRange: [0, 100], outputRange: [startAngle, endAngle] }); };
    var sumOfDonutItemValue = useMemo(function () {
        return data
            .map(function (d) { return d.value; })
            .reduce(function (total, prev) { return total + prev; });
    }, [data]);
    var donutItemValueToPercentage = useMemo(function () { return data.map(function (d) { return (d.value / sumOfDonutItemValue) * 100; }); }, [sumOfDonutItemValue, data]);
    useMemo(function () {
        var rotationRange = [];
        data.forEach(function (_, idx) {
            var fromValues = sum(donutItemValueToPercentage.slice(0, idx));
            var toValues = sum(donutItemValueToPercentage.slice(0, idx + 1));
            animatedPaths.push(new Animated.Value(LinearInterpolation(__assign({ value: fromValues }, defaultInterpolateConfig()))));
            rotationRange[idx] = {
                from: LinearInterpolation(__assign({ value: fromValues }, defaultInterpolateConfig())),
                to: LinearInterpolation(__assign({ value: toValues }, defaultInterpolateConfig())),
            };
        });
        setRotationPath(rotationRange);
    }, [data]);
    useEffect(function () {
        switch (animationType) {
            case "slide":
                animateContainerOpacity.setValue(1);
                slideAnimation();
                break;
            default:
                fadeAnimation();
                break;
        }
    }, []);
    var slideAnimation = function () {
        var animations = data.map(function (_, i) {
            var ani = Animated.timing(animatedPaths[i], {
                toValue: rotationPaths[i].to,
                duration: 3000,
                easing: Easing.bezier(0.075, 0.82, 0.165, 1),
                useNativeDriver: true,
            });
            return ani;
        });
        Animated.parallel(animations).start();
    };
    var fadeAnimation = function () {
        Animated.timing(animateContainerOpacity, {
            toValue: 1,
            duration: 5000,
            easing: Easing.bezier(0.075, 0.82, 0.165, 1),
            useNativeDriver: true,
        }).start();
    };
    useEffect(function () {
        data.forEach(function (_, i) {
            var element = pathRefs.current[i];
            donutItemListeners[i] = addListener({
                element: element,
                animatedValue: animatedPaths[i],
                startValue: rotationPaths[i].from,
            });
        });
    }, []);
    useEffect(function () {
        return function () {
            if (animationType === "slide") {
                data.forEach(function (_, i) {
                    var _a, _b;
                    if ((donutItemListeners === null || donutItemListeners === void 0 ? void 0 : donutItemListeners[i]) &&
                        (donutItemListeners === null || donutItemListeners === void 0 ? void 0 : donutItemListeners[i].removeAllListeners)) {
                        (_b = donutItemListeners === null || donutItemListeners === void 0 ? void 0 : (_a = donutItemListeners[i]).removeAllListeners) === null || _b === void 0 ? void 0 : _b.call(_a);
                    }
                });
            }
        };
    }, []);
    var addListener = function (_a) {
        var element = _a.element, animatedValue = _a.animatedValue, startValue = _a.startValue;
        animatedValue.addListener(function (angle) {
            var arcParams = {
                coordX: viewBox.getCenterCoord().x,
                coordY: viewBox.getCenterCoord().y,
                radius: radius,
                startAngle: startValue,
                endAngle: angle.value,
            };
            var drawPath = new Arc(arcParams).getDrawPath();
            if (element) {
                element.setNativeProps({ d: drawPath });
            }
        });
    };
    useEffect(function () {
        animateOpacity.setValue(0);
        Animated.timing(animateOpacity, {
            toValue: 1,
            duration: 500,
            easing: Easing.bezier(0.075, 0.82, 0.165, 1),
            useNativeDriver: true,
        }).start();
    }, []);
    var onUpdateDisplayValue = function (value, index) {
        setDisplayValue(value);
        animateOpacity.setValue(0);
        Animated.parallel([
            Animated.timing(animateOpacity, {
                toValue: 1,
                duration: 500,
                useNativeDriver: true,
            }),
        ]).start();
    };
    var onPressIn = function (value, index) {
        Animated.timing(animatedStrokeWidths[index], {
            toValue: strokeWidth + 2,
            duration: 500,
            useNativeDriver: true,
            easing: Easing.bezier(0.075, 0.82, 0.165, 1),
        }).start();
    };
    var onPressOut = function (index) {
        Animated.timing(animatedStrokeWidths[index], {
            toValue: strokeWidth,
            duration: 500,
            useNativeDriver: true,
            easing: Easing.bezier(0.075, 0.82, 0.165, 1),
        }).start();
    };
    var _getContainerStyle = function () { return [
        styles.defaultContainer,
        containerStyle,
        { width: containerWidth, height: containerHeight },
    ]; };
    var _getLabelValueStyle = function (color) { return [
        styles.defaultLabelValue,
        { color: color },
        labelValueStyle,
    ]; };
    var _getLabelTitleStyle = function (color) { return [
        styles.defaultLabelTitle,
        { color: color },
        labelTitleStyle,
    ]; };
    var _getLabelWrapperStyle = function () { return [
        styles.defaultLabelWrapper,
        {
            width: squareInCircle.getCorner() - strokeWidth,
            height: squareInCircle.getCorner() - strokeWidth,
            opacity: animateOpacity,
        },
        labelWrapperStyle,
    ]; };
    return (<Fragment>
      <View style={_getContainerStyle()}>
        <Svg width={viewBox.width} height={viewBox.height}>
          {rotationPaths.map(function (d, i) {
            var arcParams = {
                coordX: viewBox.getCenterCoord().x,
                coordY: viewBox.getCenterCoord().y,
                radius: radius,
                startAngle: d.from,
                endAngle: d.to,
            };
            var drawPath = new Arc(arcParams).getDrawPath();
            return (<AnimatedPath key={"item-".concat(i)} ref={function (el) { return (pathRefs.current[i] = el); }} onPress={function () { return onUpdateDisplayValue(data[i], i); }} onPressIn={function () { return onPressIn(data[i], i); }} onPressOut={function () { return onPressOut(i); }} strokeLinecap={type} d={drawPath} opacity={animateContainerOpacity} fill="none" stroke={data[i].color} strokeWidth={animatedStrokeWidths[i]}/>);
        })}
        </Svg>
        <Animated.View style={_getLabelWrapperStyle()}>
          <Text style={_getLabelValueStyle(displayValue === null || displayValue === void 0 ? void 0 : displayValue.color)}>
            {displayValue === null || displayValue === void 0 ? void 0 : displayValue.value}
          </Text>
          <Text style={_getLabelTitleStyle(displayValue === null || displayValue === void 0 ? void 0 : displayValue.color)}>
            {displayValue === null || displayValue === void 0 ? void 0 : displayValue.name}
          </Text>
        </Animated.View>
      </View>
    </Fragment>);
};
var styles = StyleSheet.create({
    defaultContainer: {
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
    },
    defaultLabelWrapper: {
        position: "absolute",
        justifyContent: "center",
        alignItems: "center",
    },
    defaultLabelValue: {
        fontSize: 32,
        fontWeight: "bold",
    },
    defaultLabelTitle: {
        fontSize: 16,
    },
});
